### PR TITLE
Fix #24702: Voice too long corruption

### LINKED ISSUE #24702: Voice too long corruption
### Issue type

File corruption

### Description with steps to reproduce

Since 4.4.0 it is possible to make voice 2 longer than actual measure size. See video
During investigation for [this report ](https://musescore.org/en/node/368654) I found one way to do this. Although I don’t know what exactly caused the corruption in that score.
This is a regression since 4.4.0. 
Can't reproduce in 4.3.2

### Supporting files, videos and screenshots

https://github.com/user-attachments/assets/19666a81-2d31-4194-abed-d608058f551d

### What is the latest version of MuseScore Studio where this issue is present?

MuseScore Studio version (64-bit): 4.5.0-242560303, revision: github-musescore-musescore-ac76040

### Regression

Yes, this used to work in a previous version of MuseScore 4.x

### Operating system

windows 10

### Additional context

_No response_

### Checklist

- [X] This report follows the [guidelines](https://github.com/musescore/MuseScore/wiki/Reporting-bugs-and-issues) for reporting bugs and issues
- [X] I have verified that this issue has not been logged before, by searching the [issue tracker](https://github.com/musescore/MuseScore/issues) for similar issues
- [X] I have attached all requested files and information to this report
- [X] I have attempted to identify the root problem as concisely as possible, and have used minimal reproducible examples where possible

### PR DESCRIPTION
**Pull request by mathesoncalum**
Resolves: #24702 

Introduced with #23911

The idea here is to scan backwards from the input position to find the previous `ChordRest`, check if the duration overlaps with the input position, shorten the previous `ChordRest` by the length of the overlap, and fill any remaining gaps with tied notes (if necessary).

This PR also addresses some janky input when switching voices in the middle of a tuplet (shown below) - for now we'll simply move the input cursor to the beginning of the tuplet. In future we may consider adding tuplets to the "new" voice in this case, but a few UX considerations will need to be resolved before we make this happen.

https://github.com/user-attachments/assets/7b326df7-3cf3-4fcc-9e9f-9fd49cf6249e

### CODE REVIEW

**Comment by cbjeukendrup** [comment]
It works well when the new duration of the previous note can be represented as a single duration, but otherwise still causes corruptions:

https://github.com/user-attachments/assets/8d0ea5ef-4207-4fa8-b3d4-089bb551401c

So `changeCrLen` is unable to split a note into multiple tied notes when it cannot be represented as a single note. There must be some other method somewhere in the engraving module that is able to do this, but I wouldn't know where.

**Comment by mathesoncalum** [comment]
@cbjeukendrup apparently it _can_ actually split the note into multiple ties, but only if the target `ChordRest` is being lengthened instead of shortened... Looking into it.

File: src/engraving/dom/noteentry.cpp [hunk 1]
```diff
@@ -450,6 +450,34 @@ Ret Score::putNote(const Position& p, bool replace)
...
+            const Fraction desiredDuration = prevCr->ticks() - overlapDuration;
+
+            const InputState inputStateToRestore = m_is; // because changeCRlen will alter the input state
+            ms->changeCRlen(prevCr, desiredDuration, /*fillWithRest*/ false);
+
+            // Fill the difference with tied notes if necessary...
+            const Fraction difference = desiredDuration - prevCr->ticks();
+            if (prevCr->isChord() && difference.isNotZero()) {
+                Fraction startTick = prevCr->endTick();
+                Chord* prevChord = nullptr;
```
Comment (cbjeukendrup) [inline src/engraving/dom/noteentry.cpp:468]: Wouldn't it be slightly simpler to do
```suggestion
                Chord* prevChord = toChord(prevCr);
```
? Then you don't need the `tieFrom` variable.

**Review by cbjeukendrup — APPROVED** [review]

**Review by cbjeukendrup — APPROVED** [review]

**Review by oktophonie — APPROVED** [review]

**Comment by zacjansheski** [comment]
Tested on MacOS 14, Windows 11, Ubuntu 22.04.3. Approved
https://github.com/musescore/MuseScore/issues/24702

**Pull request merged**